src/treewidth.jl

```diff
@@ -64,149 +64,190 @@ The `BT` algorithm is an exact solver for the treewidth problem that implemented
-        if length(vs) >= 2
-            sub_list_indices = unique!(vcat([incidence_list.v2e[v] for v in vs]...)) # the vertices connected to the tensors to be contracted
-            sub_list_open_indices = setdiff(sub_list_indices, eliminated_vertices) # the vertices connected to the tensors to be contracted but not eliminated
-            vmap = Dict([i => incidence_list.v2e[v] for (i, v) in enumerate(vs)])
-            sub_list = IncidenceList(vmap; openedges=sub_list_open_indices) # the subgraph of the contracted tensors
-            sub_tree, scs, tcs = tree_greedy(sub_list, log2_edge_sizes; α=0.0, temperature=0.0) # optmize the subgraph with greedy method
-            sub_tree = expand_indices(sub_tree, Dict([i => v for (i, v) in enumerate(vs)]))
-            vi = contract_tree!(incidence_list, sub_tree, log2_edge_sizes, scs, tcs) # insert the contracted tensors back to the total graph
-            contraction_tree_nodes[tensors_list[vi]] = st2ct(sub_tree, tensors_list, contraction_tree_nodes)
-            flag = vi
+        for e in sep
+            push!(code.eins.iy, el[e])
         end
-    end
 
-    return contraction_tree_nodes[tensors_list[flag]]
-end
+        for e in res, v in view(rowvals(ve), nzrange(ve, e))
+            if marker[v] != -2
+                marker[v] = -2
 
-function expand_indices(sub_tree::Union{ContractionTree, VT}, vmap::Dict{Int, VT}) where{VT}
-    if sub_tree isa ContractionTree
-        return ContractionTree(expand_indices(sub_tree.left, vmap), expand_indices(sub_tree.right, vmap))
-    else
-        return vmap[sub_tree]
+                if v == n
+                    append!(code.eins.iy, iy)
+                else
+                    push!(code.args, NestedEinsum{L}(v))
+                    push!(code.eins.ixs, ixs[v])
+                end
+            end
+        end
+
+        for _ in childindices(tree, b)
+            child = pop!(stack)
+            push!(code.args, child)
+            push!(code.eins.ixs, child.eins.iy)
+        end
+
+        push!(stack, code)
     end
-end
 
-function st2ct(sub_tree::Union{ContractionTree, VT}, tensors_list::Dict{VT, Int}, contraction_tree_nodes::Vector) where{VT}
-    if sub_tree isa ContractionTree
-        return ContractionTree(st2ct(sub_tree.left, tensors_list, contraction_tree_nodes), st2ct(sub_tree.right, tensors_list, contraction_tree_nodes))
+    # we now have an expression for each root
+    # of the tree decomposition
+    if isone(length(stack))
+        code = only(stack)
     else
-        return contraction_tree_nodes[tensors_list[sub_tree]]        
-    end
-end
+        code = NestedEinsum(NestedEinsum{L}[], EinCode(Vector{L}[], L[]))
+        append!(code.eins.iy, iy)
 
-"""
-    optimize_treewidth(optimizer, eincode, size_dict)
+        while !isempty(stack)
```

Comment:
I think it is fine to construct a non-binary tree like this and then use `GreedyMethod` to optimize based on it. Simple code is more robust. If there is a performance issue, I can improve the implementation of `GreedyMethod`.